resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala

```diff
@@ -0,0 +1,161 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.deploy.k8s.features
+
+import scala.collection.JavaConverters._
+
+import com.google.common.base.Charsets
+import com.google.common.io.Files
+import io.fabric8.kubernetes.api.model.{ConfigMapBuilder, HasMetadata}
+
+import org.apache.spark.deploy.SparkHadoopUtil
+import org.apache.spark.deploy.k8s.{KubernetesConf, KubernetesUtils, SparkPod}
+import org.apache.spark.deploy.k8s.Config._
+import org.apache.spark.deploy.k8s.Constants._
+import org.apache.spark.deploy.k8s.KubernetesDriverSpecificConf
+import org.apache.spark.deploy.k8s.features.hadooputils._
+import org.apache.spark.internal.Logging
+
+ /**
+  * Runs the necessary Hadoop-based logic based on Kerberos configs and the presence of the
+  * HADOOP_CONF_DIR. This runs various bootstrap methods defined in HadoopBootstrapUtil.
+  */
+private[spark] class KerberosConfDriverFeatureStep(
+    kubernetesConf: KubernetesConf[KubernetesDriverSpecificConf])
+    extends KubernetesFeatureConfigStep with Logging {
+
+    private val conf = kubernetesConf.sparkConf
+    private val maybePrincipal = conf.get(org.apache.spark.internal.config.PRINCIPAL)
+    private val maybeKeytab = conf.get(org.apache.spark.internal.config.KEYTAB)
+    private val maybeExistingSecretName = conf.get(KUBERNETES_KERBEROS_DT_SECRET_NAME)
+    private val maybeExistingSecretItemKey =
+      conf.get(KUBERNETES_KERBEROS_DT_SECRET_ITEM_KEY)
+    private val maybeKrb5File =
+      conf.get(KUBERNETES_KERBEROS_KRB5_FILE)
+    private val maybeKrb5CMap =
+      conf.get(KUBERNETES_KERBEROS_KRB5_CONFIG_MAP)
+    private val kubeTokenManager = kubernetesConf.tokenManager(conf,
+      SparkHadoopUtil.get.newConfiguration(conf))
+    private val isKerberosEnabled = kubeTokenManager.isSecurityEnabled
+
+    require(maybeKeytab.isEmpty || isKerberosEnabled,
+      "You must enable Kerberos support if you are specifying a Kerberos Keytab")
+
+    require(maybeExistingSecretName.isEmpty || isKerberosEnabled,
+      "You must enable Kerberos support if you are specifying a Kerberos Secret")
+
+   require((maybeKrb5File.isEmpty || maybeKrb5CMap.isEmpty) || isKerberosEnabled,
+    "You must specify either a krb5 file location or a ConfigMap with a krb5 file")
+
+   KubernetesUtils.requireNandDefined(
```

Comment:
Does the check guarantee that one or the other must be specified? It looks to me that `requireNandDefined` only guarantees that when the first is specified, the second must be empty.

Reply:
I didn't think it was necessary since it was a one-directional issue. But I reformatted the check to be bi-directional.